File: packages/adapters/ios/AwakeEventCollector/RawEventBuffer.swift
import Foundation

/// Thread-safe bounded FIFO of RawEvents observed natively but not yet handed to
/// JS. On overflow the OLDEST entry is dropped. Backstop only — the module also
/// pushes batches as signals arrive.
final class RawEventBuffer {
    private var items: [RawEvent] = []
    private let capacity: Int
    private let queue = DispatchQueue(label: "os.awake.collector.buffer")

    init(capacity: Int = 4096) { self.capacity = capacity }

    func add(_ event: RawEvent) {
        queue.sync {
            if items.count >= capacity { items.removeFirst(items.count - capacity + 1) }
            items.append(event)
        }
    }

    func drain() -> [RawEvent] {
        queue.sync {
            let out = items
            items.removeAll(keepingCapacity: true)
            return out
        }
    }

    func size() -> Int { queue.sync { items.count } }
}

File: apps/mobile/android/app/src/main/java/os/awake/collector/RawEventBuffer.kt
package os.awake.collector

import java.util.ArrayDeque

/**
 * Thread-safe bounded FIFO of RawEvents observed on the native side but not yet
 * handed to JS. Drains via `drainPendingEvents()`. On overflow the OLDEST entry
 * is dropped (never the newest) so a burst never loses the most recent signal.
 *
 * This is only a backstop for signals observed while the JS side was asleep; the
 * module also pushes batches as they arrive.
 */
class RawEventBuffer(private val capacity: Int = 4096) {

    private val queue = ArrayDeque<RawEvent>()
    private val lock = Any()

    fun add(event: RawEvent) {
        synchronized(lock) {
            while (queue.size >= capacity) queue.pollFirst()
            queue.addLast(event)
        }
    }

    fun drain(): List<RawEvent> = synchronized(lock) {
        val out = ArrayList<RawEvent>(queue.size)
        out.addAll(queue)
        queue.clear()
        out
    }

    fun size(): Int = synchronized(lock) { queue.size }
}
